src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt

```diff
@@ -3,49 +3,147 @@ package net.atos.zac.smartdocuments.templates
+        id = smartDocumentsTemplateGroup.id,
+        name = smartDocumentsTemplateGroup.name,
+        groups = null,
+        templates = null
+    )
+
+    private fun createRESTTemplate(
+        smartDocumentsTemplate: SmartDocumentsResponseTemplate
+    ) = RESTSmartDocumentsTemplate(
+        id = smartDocumentsTemplate.id,
+        name = smartDocumentsTemplate.name
+    )
+
+    fun Set<RESTSmartDocumentsTemplateGroup>.toModel(
+        zaakafhandelParameters: ZaakafhandelParameters
+    ): Set<SmartDocumentsTemplateGroup> =
+        this.mapTo(mutableSetOf()) { convertTemplateGroupToModel(it, null, zaakafhandelParameters) }
 
     private fun convertTemplateGroupToModel(
-        group: SmartDocumentsResponseTemplateGroup,
-        parent: SmartDocumentsTemplateGroup?
+        group: RESTSmartDocumentsTemplateGroup,
+        parent: SmartDocumentsTemplateGroup?,
+        zaakafhandelParameterId: ZaakafhandelParameters
     ): SmartDocumentsTemplateGroup {
-        val jpaGroup = createModelTemplateGroup(group, parent)
+        val jpaGroup = createModelTemplateGroup(group, parent, zaakafhandelParameterId)
 
         group.templates?.forEach {
-            jpaGroup.templates.add(createModelTemplate(it))
+            jpaGroup.templates.add(createModelTemplate(it, jpaGroup, zaakafhandelParameterId))
         }
 
-        group.templateGroups?.forEach {
-            jpaGroup.children.add(convertTemplateGroupToModel(it, jpaGroup))
+        group.groups?.forEach {
+            jpaGroup.children.add(convertTemplateGroupToModel(it, jpaGroup, zaakafhandelParameterId))
         }
 
         return jpaGroup
     }
 
     private fun createModelTemplateGroup(
-        smartDocumentsTemplateGroup: SmartDocumentsResponseTemplateGroup,
-        parentGroup: SmartDocumentsTemplateGroup?
+        smartDocumentsTemplateGroup: RESTSmartDocumentsTemplateGroup,
+        parentGroup: SmartDocumentsTemplateGroup?,
+        zaakafhandelParams: ZaakafhandelParameters
     ) = SmartDocumentsTemplateGroup().apply {
         smartDocumentsId = smartDocumentsTemplateGroup.id
+        zaakafhandelParameters = zaakafhandelParams
         name = smartDocumentsTemplateGroup.name
         parent = parentGroup
         creationDate = ZonedDateTime.now()
     }
 
     private fun createModelTemplate(
-        smartDocumentsTemplate: SmartDocumentsResponseTemplate
+        smartDocumentsTemplate: RESTSmartDocumentsTemplate,
+        parentGroup: SmartDocumentsTemplateGroup,
+        zaakafhandelParams: ZaakafhandelParameters
     ) = SmartDocumentsTemplate().apply {
         smartDocumentsId = smartDocumentsTemplate.id
+        zaakafhandelParameters = zaakafhandelParams
         name = smartDocumentsTemplate.name
+        templateGroup = parentGroup
         creationDate = ZonedDateTime.now()
     }
+
+    fun Set<SmartDocumentsTemplateGroup>.toREST(): Set<RESTSmartDocumentsTemplateGroup> =
+        this.mapTo(mutableSetOf()) { convertTemplateGroupToREST(it) }
```

Comment:
I am not sure why we need these `mutableSet`s, these are just model classes right? We're not going to make any changes to the sets nor contained data, correct?

Reply:
This comes from the JPA model. ManyToOne and OneToMany cannot be sets, but are required to be mutable

Reply:
This one in particular is the start of the recursion, so `convertTemplateGroupToREST` adds data to the set.